### PR TITLE
Support for python 3.11 and 3.12 [GEOPY-1494]

### PR DESCRIPTION
**Pull request by SamuelBradley**
**GEOPY-1494 - Python 3.11 (and 3.12) Support**
Made the following changes
- Update Poetry file to allow python version <4 (Good practise to not arbitrarily restrict use with new minor python versions)
- Update lock file
- Run tests again python 3.12.1
- Update Github actions to run tests in 3.11 and 3.12

### CODE REVIEW

**Comment by domfournier** [comment]
Hi @SamuelBradley 
Very sorry we took a while to get back to you on this.

Would you mind updating with the current state of develop? Then request my review, we can bring this in on the next release.

Thanks for contributing!

**Comment by SamuelBradley** [comment]
@domfournier, Thanks for the response Aligned this branch to the Head of develop, retested in python 3.11 and 3.12 and added a test fix.

File: .github/workflows/static-analysis.yml
Comment (domfournier) [inline .github/workflows/static-analysis.yml:1]: Sorry @SamuelBradley . Couple more changes requested by @sebhmg on my mock PR

https://github.com/MiraGeoscience/geoh5py/pull/561#issuecomment-2150963538

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -39,25 +40,24 @@ repos:
     rev: v3.15.2
     hooks:
     -   id: pyupgrade
-        args: [--py37-plus]
+        args: [--py39-plus]
```
Comment (domfournier) [inline .pre-commit-config.yaml:43]: Also here

```suggestion
        args: [--py310-plus]
```

**Review by domfournier — COMMENTED** [review]
Couple more from a previous PR. We had to abandon 3.9 to resolve conflicts with scipy.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (SamuelBradley) [reply .pre-commit-config.yaml]: @domfournier I have committed changes including the requested line which removes support for python 3.9. Changes pass the pre-commit checks and tests are still passing in python 3.12

**Review by domfournier — APPROVED** [review]

File: pyproject.toml [hunk 2]
```diff
@@ -39,10 +39,12 @@ include = [
...
 
 [tool.poetry.dependencies]
-python = "^3.9,<3.11"
-
+python = "^3.10,<4"
 h5py = "^3.2.1"
-numpy = "!=1.19.4, ~1.23.5"
+numpy = [
+    {version="!=1.19.4, ^1.23.5", python= "<3.12"},
+    {version="^1.26.0", python= ">=3.12"}
```
Comment (sebhmg) [inline pyproject.toml:46]: numpy 1.26 seems to support Python versions from 3.9 to 3.12, so the same version could be used no matter the version of Python

(from [numpy-1-26-0 release notes](https://numpy.org/doc/1.26/release/1.26.0-notes.html#numpy-1-26-0-release-notes))

File: pyproject.toml [hunk 3]
```diff
@@ -52,7 +54,10 @@ lockfile = "^0.12.2"
 pylint = "*"
 pytest = "*"
 pytest-cov = "*"
-scipy = "~1.10.1"
+scipy = [
+    {version="~1.10.1", python= "<3.12"},
+    {version=">=1.12.0", python= ">=3.12"}
```
Comment (sebhmg) [inline pyproject.toml:59]: if possible, just always use SciPy 1.12, not matter what the version of Python
(according to [SciPy 1.12.0 release notes](https://docs.scipy.org/doc/scipy/release/1.12.0-notes.html#scipy-1-12-0-release-notes), SciPy 1.12 supports Python 3.9+)

**Review by sebhmg — APPROVED** [review]

File: .github/workflows/static-analysis.yml [hunk 4]
```diff
@@ -40,7 +40,7 @@ jobs:
       - name: Set up Python version
         uses: actions/setup-python@v5
         with:
-          python-version: 3.10
+          python-version: "3.10"
```
Comment (domfournier) [inline .github/workflows/static-analysis.yml:43]: @sebhmg To remember when we switch over the other repos ...

**Review by domfournier — APPROVED** [review]

**Review by sebhmg — APPROVED** [review]

**Comment by domfournier** [comment]
@andrewg-mira We will need your third-party approval here since we all pushed something...

File: pyproject.toml (hunk 3 above)
Comment (sebhmg) [reply pyproject.toml:59]: ok, we can take care of it later

**Review by sebhmg — APPROVED** [review]

**Comment by domfournier** [comment]
Thank you @SamuelBradley. Your contribution is much appreciated.

**Pull request merged**